stat/combin/combin.go

```diff
@@ -284,23 +284,67 @@ func IndexToCombination(dst []int, idx, n, k int) []int {
 //  [ 1 2 0 ]
 // Cartesian panics if any of the provided lengths are less than 1.
 func Cartesian(lens []int) [][]int {
-	if len(lens) == 0 {
+	rows := Card(lens)
+	if rows == 0 {
 		panic("combin: empty lengths")
 	}
-	rows := 1
-	for _, v := range lens {
-		if v < 1 {
-			panic("combin: length less than zero")
-		}
-		rows *= v
-	}
 	out := make([][]int, rows)
 	for i := 0; i < rows; i++ {
 		out[i] = SubFor(nil, i, lens)
 	}
 	return out
 }
 
+// Card computes the cardinality of the multi-dimensional space whose dimensions have size specified by dims
+// All length values must be positive, otherwise this will panic.
+func Card(dims []int) int {
+	if len(dims) == 0 {
+		return 0
+	}
+	card := 1
+	for _, v := range dims {
+		if v < 1 {
+			panic("combin: length less than zero")
+		}
+		card *= v
+	}
+	return card
+}
+
+// NewCartesianGenerator returns a CartesianGenerator for iterating over cartesian products which are generated on the fly.
+// At least one length must be provided and all length values must be positive, otherwise this will panic.
```

Comment:
This doesn't panic if one length isn't provided. I think for now add the same check in Product.

Reply:
solved with 5a4529e